File: src/main/java/dev/gabrielchl/intellijpets/toolWindow/Pet.java
package dev.gabrielchl.intellijpets.toolWindow;

import com.intellij.util.ui.JBUI;
import dev.gabrielchl.intellijpets.settings.AppSettingsState;
import dev.gabrielchl.intellijpets.utils.Utils;

import javax.swing.*;
import java.awt.*;
import java.awt.event.MouseAdapter;
import java.awt.event.MouseEvent;
import java.awt.event.MouseMotionAdapter;
import java.awt.geom.AffineTransform;
import java.awt.image.*;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Random;

public class Pet {
    private enum State {
        ATTACK,
        IDLE,
        JUMP,
        LIEDOWN,
        RUN,
        SIT,
        WALK,
    }

    private final JPanel container;
    private final JPanel petInnerPanel;
    private final JLabel pet;
    private int xIndex;
    private State state;
    private int currentX;
    private int targetX;
    private final HashMap<State, ArrayList<Image>> sprites = new HashMap<>();
    private final int SPRITE_WIDTH;
    private final int SPRITE_HEIGHT;
    private final int SPEED;

    public Pet(JPanel container, JPanel petInnerPanel, JLabel pet) {
        this.container = container;
        this.petInnerPanel = petInnerPanel;
        this.pet = pet;

        xIndex = 0;
        state = State.SIT;
        targetX = 0;

        AppSettingsState settings = AppSettingsState.getInstance().getState();
        String petVariant = settings == null ? "cat-1" : settings.getPetVariant();

        int spriteSize = Utils.petToSpriteSize.get(petVariant);
        double display_scale = 1.6;
        double spriteWidthBase = Math.round(spriteSize * display_scale);
        double spriteHeightBase = Math.round(spriteSize * display_scale);
        double speed_base = 7d / 40 * spriteSize;

        double petScale = settings == null ? 1d : settings.getPetScale();
        SPRITE_WIDTH = (int)Math.round(spriteWidthBase * petScale);
        SPRITE_HEIGHT = (int)Math.round(spriteHeightBase * petScale);
        SPEED = (int)Math.round(speed_base * petScale);

        for (State state : State.values()) {
            Image spriteImg = Toolkit.getDefaultToolkit().createImage(getClass().getResource(String.format("/spritesheets/%s/%s.png", petVariant, state.toString().toLowerCase())));
            Utils.waitForImageToLoad(spriteImg);
            ImageProducer catSource = spriteImg.getSource();

            ArrayList<Image> spriteRow = new ArrayList<>();
            for (int x = 0; x < spriteImg.getWidth(null) / spriteSize; x += 1) {
                Image img = Toolkit.getDefaultToolkit().createImage(
                        new FilteredImageSource(
                                catSource,
                                new CropImageFilter(spriteSize * x, 0, spriteSize, spriteSize)
                        )
                ).getScaledInstance(SPRITE_WIDTH, SPRITE_HEIGHT, Image.SCALE_DEFAULT);
                Utils.waitForImageToLoad(img);
                spriteRow.add(img);
            }
            sprites.put(state, spriteRow);
        }

        container.addMouseListener(new MouseAdapter() {
            @Override
            public void mouseClicked(MouseEvent e) {
                super.mouseClicked(e);
                if (state == State.SIT) {
                    xIndex = 0;
                    state = Math.round((float)Math.random()) == 0 ? State.JUMP : State.ATTACK;
                }
            }
        });
        container.addMouseMotionListener(new MouseMotionAdapter() {
            @Override
            public void mouseMoved(MouseEvent e) {
                super.mouseMoved(e);
                if (Math.abs(e.getX() - (currentX + (SPRITE_WIDTH / 2))) > 30) {
                    targetX = Math.max(0, e.getX() - (SPRITE_WIDTH / 2));
                }
            }
        });
        container.addMouseListener(new MouseAdapter() {
            @Override
            public void mouseExited(MouseEvent e) {
                super.mouseExited(e);
                targetX = currentX;
            }
        });
    }

    public void tick() {
        if (currentX != targetX && state != State.WALK) {
            state = State.WALK;
            xIndex = 0;
        }
        if (currentX == targetX && state == State.WALK) {
            state = State.SIT;
            xIndex = 0;
        }
        if (xIndex >= sprites.get(state).size()) {
            xIndex = 0;
            if (state == State.JUMP || state == State.ATTACK) {
                state = State.SIT;
            } else if (state == State.SIT) {
                int rand = new Random().nextInt(10);
                if (rand == 0) {
                    state = State.IDLE;
                }
                if (rand == 1) {
                    state = State.LIEDOWN;
                }
                if (rand == 2) {
                    if (container.getWidth() > 0) {
                        targetX = new Random().nextInt(container.getWidth() - SPRITE_WIDTH);
                    }
                }
            } else if (state == State.IDLE) {
                int rand = new Random().nextInt(10);
                if (rand == 0) {
                    state = State.SIT;
                }
                if (rand == 1) {
                    if (container.getWidth() > 0) {
                        targetX = new Random().nextInt(container.getWidth() - SPRITE_WIDTH);
                    }
                }
            } else if (state == State.LIEDOWN) {
                if (new Random().nextInt(15) == 0) {
                    state = State.SIT;
                }
            }
        }
        boolean toRight = true;
        if (currentX != targetX) {
            int change = Math.abs(targetX - currentX);
            if (change > SPEED) {
                change = SPEED;
            }
            if (targetX < currentX) {
                change *= -1;
                toRight = false;
            }
            currentX = currentX + change;
            petInnerPanel.setBorder(JBUI.Borders.emptyLeft(currentX));
        }
        Image catImg = sprites.get(state).get(xIndex);
        if (!toRight) {
            BufferedImage originalImage = Utils.toBufferedImage(catImg);
            AffineTransform tx = AffineTransform.getScaleInstance(-1, 1);
            tx.translate(-originalImage.getWidth(null), 0);
            AffineTransformOp op = new AffineTransformOp(tx, AffineTransformOp.TYPE_NEAREST_NEIGHBOR);
            BufferedImage flippedImage = op.filter(originalImage, null);
            ImageIcon catIcon = new ImageIcon(flippedImage.getScaledInstance(SPRITE_WIDTH, SPRITE_HEIGHT, Image.SCALE_DEFAULT));
            pet.setIcon(catIcon);
        } else {
            ImageIcon catIcon = new ImageIcon(catImg);
            pet.setIcon(catIcon);
        }
        xIndex = xIndex + 1;
    }
}

File: src/main/kotlin/dev/gabrielchl/intellijPets/toolWindow/Pet.kt
package dev.gabrielchl.intellijPets.toolWindow

import dev.gabrielchl.intellijPets.settings.PetsSettings
import dev.gabrielchl.intellijPets.utils.Constants
import dev.gabrielchl.intellijPets.utils.Utilities
import java.awt.Image
import java.awt.Toolkit
import java.awt.event.MouseEvent
import java.awt.geom.AffineTransform
import java.awt.image.AffineTransformOp
import java.awt.image.BufferedImage
import java.awt.image.CropImageFilter
import java.awt.image.FilteredImageSource
import java.util.*
import javax.swing.JPanel
import kotlin.math.abs
import kotlin.math.max

// TODO: rewrite all this to allow more pet-specific behaviours
class Pet(val variant: String, val container: JPanel) {
    private enum class State {
        ATTACK,
        IDLE,
        JUMP,
        LIEDOWN,
        RUN,
        SIT,
        WALK,
    }

    private var xIndex = 0
    private var state: State
    var currentX = Random().nextInt(300)
    private var targetX = currentX
    private val sprites: HashMap<State, ArrayList<Image>> = HashMap<State, ArrayList<Image>>()
    val SPRITE_WIDTH: Int
    val SPRITE_HEIGHT: Int
    private val SPEED: Int
    var image: Image

    init {
        state = State.SIT

        val settings = PetsSettings.instance.state

        val spriteSize: Int = Constants.PET_TO_SPRITE_SIZE[variant]!!
        val displayScale = 1.6
        val spriteWidthBase = Math.round(spriteSize * displayScale).toDouble()
        val spriteHeightBase = Math.round(spriteSize * displayScale).toDouble()
        val speedBase = 7.0 / 40 * spriteSize

        SPRITE_WIDTH = Math.round(spriteWidthBase * settings.petScale).toInt()
        SPRITE_HEIGHT = Math.round(spriteHeightBase * settings.petScale).toInt()
        SPEED = Math.round(speedBase * settings.petScale).toInt()

        for (state in State.values()) {
            val spriteImg = Toolkit.getDefaultToolkit().createImage(
                javaClass.getResource(
                    String.format(
                        "/spritesheets/%s/%s.png", variant, state.toString().lowercase(
                            Locale.getDefault()
                        )
                    )
                )
            )
            Utilities.waitForImageToLoad(spriteImg)
            val catSource = spriteImg.source

            val spriteRow = ArrayList<Image>()
            var x = 0
            while (x < spriteImg.getWidth(null) / spriteSize) {
                val img = Toolkit.getDefaultToolkit().createImage(
                    FilteredImageSource(
                        catSource,
                        CropImageFilter(spriteSize * x, 0, spriteSize, spriteSize)
                    )
                ).getScaledInstance(SPRITE_WIDTH, SPRITE_HEIGHT, Image.SCALE_DEFAULT)
                Utilities.waitForImageToLoad(img)
                spriteRow.add(img)
                x += 1
            }
            sprites[state] = spriteRow
        }
        image = sprites[state]!![xIndex]
    }

    fun onMouseClicked(e: MouseEvent) {
        if (state == Pet.State.SIT) {
            xIndex = 0
            state = if (Math.round(Math.random().toFloat()) == 0) Pet.State.JUMP else Pet.State.ATTACK
        }
    }

    fun onMouseMoved(e: MouseEvent) {
        if (abs((e.x - (currentX + (SPRITE_WIDTH / 2))).toDouble()) > 30) {
            targetX = max(0.0, (e.x - (SPRITE_WIDTH / 2)).toDouble()).toInt()
        }
    }

    fun onMouseExited(e: MouseEvent) {
        if (state == Pet.State.SIT) {
            targetX = currentX
        }
    }

    fun tick() {
        if (currentX != targetX && state != Pet.State.WALK) {
            state = Pet.State.WALK
            xIndex = 0
        }
        if (currentX == targetX && state == Pet.State.WALK) {
            state = Pet.State.SIT
            xIndex = 0
        }
        if (xIndex >= sprites[state]!!.size) {
            xIndex = 0
            if (state == Pet.State.JUMP || state == Pet.State.ATTACK) {
                state = Pet.State.SIT
            } else if (state == Pet.State.SIT) {
                val rand = Random().nextInt(10)
                if (rand == 0) {
                    state = Pet.State.IDLE
                }
                if (rand == 1) {
                    state = Pet.State.LIEDOWN
                }
                if (rand == 2) {
                    if (container.width > 0) {
                        targetX = Random().nextInt(container.width - SPRITE_WIDTH)
                    }
                }
            } else if (state == Pet.State.IDLE) {
                val rand = Random().nextInt(10)
                if (rand == 0) {
                    state = Pet.State.SIT
                }
                if (rand == 1) {
                    if (container.width > 0) {
                        targetX = Random().nextInt(container.width - SPRITE_WIDTH)
                    }
                }
            } else if (state == Pet.State.LIEDOWN) {
                if (Random().nextInt(15) == 0) {
                    state = Pet.State.SIT
                }
            }
        }
        var toRight = true
        if (currentX != targetX) {
            var change = abs((targetX - currentX).toDouble()).toInt()
            if (change > SPEED) {
                change = SPEED
            }
            if (targetX < currentX) {
                change *= -1
                toRight = false
            }
            currentX += change
        }
        val catImg = sprites[state]!![xIndex]
        if (!toRight) {
            val originalImage: BufferedImage = Utilities.toBufferedImage(catImg)
            val tx = AffineTransform.getScaleInstance(-1.0, 1.0)
            tx.translate(-originalImage.getWidth(null).toDouble(), 0.0)
            val op = AffineTransformOp(tx, AffineTransformOp.TYPE_NEAREST_NEIGHBOR)
            val flippedImage = op.filter(originalImage, null)
            image = flippedImage.getScaledInstance(SPRITE_WIDTH, SPRITE_HEIGHT, Image.SCALE_DEFAULT)
        } else {
            image = catImg
        }
        xIndex += 1
    }
}